File: 3080-mark-elements-on-array-by-performing-queries/3080-mark-elements-on-array-by-performing-queries.cpp
class Solution {
public:
    vector<long long> unmarkedSumArray(vector<int>& nums, vector<vector<int>>& queries) {
        vector<long long> ans; // SC:O(n)
        
        unordered_map<int,bool> mp; // map -> {idx,visited}, SC:O(n)
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq; // min heap --> {val,idx}, SC:O(n)
        long long sum = 0;
        for(int i = 0; i < nums.size(); i++) { // TC:O(n) * O(logn) = O(n.logn)
            mp[i] = 0;
            pq.push({nums[i],i});
            sum += nums[i];
        }
        
        for(auto query : queries) { // TC:O(k) * O(n) = TC:O(k.n)
            int i = query[0];
            int k = query[1];
            
            // 1st task
            if(mp[i] == 0) {
                mp[i] = 1;
                sum -= nums[i];
            }
            
            // 2nd task
            while(!pq.empty() && k >= 1) { // TC:O(k.logn)
                pair<int,int> top = pq.top();
                pq.pop();
                if(mp[top.second] == false && k >= 1) {
                    // cout << top.first << " " << top.second << endl;
                    sum -= top.first;
                    mp[top.second] = 1;
                    k--;
                }
            }
            
            ans.push_back(sum);
        }
        
        return ans;
    }
};

File: 3080-mark-elements-on-array-by-performing-queries/3080-mark-elements-on-array-by-performing-queries.java
class Solution {
    public long[] unmarkedSumArray(int[] nums, int[][] queries) {
        int n = queries.length;
        long[] answer = new long[n];

        long total = 0;
        Set<Integer> markedIndices = new HashSet<>();
        
        PriorityQueue<Integer> pq = new PriorityQueue<>((i1, i2) -> {
            if (nums[i1] != nums[i2]) {
                return nums[i1] - nums[i2];
            }
            return i1 - i2;
        });
        for (int i = 0; i < nums.length; i++) {
            total += nums[i];
            pq.add(i);
        }

        for (int i = 0; i < queries.length; i++) {
            if (pq.isEmpty()) {
                break;
            }
            
            int[] q = queries[i];
            int index = q[0];
            int k = q[1];

            if (markedIndices.add(index)) {
                total -= nums[index];
            }

            while (!pq.isEmpty() && k != 0) {
                index = pq.remove();
                if (markedIndices.add(index)) {
                    k--;
                    total -= nums[index];
                }
            }
            
            answer[i] = total;
        }

        return answer;
    }
}